### PR TITLE
Add aten::grid_sampler_2d

### PR DESCRIPTION
**Pull request by majing921201**
Implement: https://github.com/pytorch/pytorch/issues/127002

### CODE REVIEW

**Review by dvrogozh — COMMENTED** [review]
I suggest you can add a link to pytorch issue to commit message:
```
Fixes: https://github.com/pytorch/pytorch/issues/127002
```

I tried on my side. Works for me.
Do you plan to add any tests to this PR?

**Comment by fengyuan14** [comment]
> I suggest you can add a link to pytorch issue [pytorch/pytorch#127002](https://github.com/pytorch/pytorch/issues/127002) to commit message:
> 
> ```
> Fixes: https://github.com/pytorch/pytorch/issues/127002
> ```
> 
> I tried on my side. Works for me. Do you plan to add any tests to this PR?

We will follow PyTorch unit test cases and their coverage. With a quick scan, https://github.com/pytorch/pytorch/blob/main/test/test_nn.py and https://github.com/pytorch/pytorch/blob/main/test/test_ops.py should be able to cover the semantics defined by PyTorch. These two cases have been ported in torch-xpu-ops. When we add the XPU implementation registration and remove CPU fallback implementation of it, the cases will test XPU implementation.

**Comment by fengyuan14** [comment]
@majing921201 Please resolve CUDA bias codes in test_nn and remove cases from skip list.
image

File: src/aten/sycl/GridSampler.h
Comment (fengyuan14) [inline src/aten/sycl/GridSampler.h:1]: Should reuse PyTorch header.
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/GridSampler.h

File: src/aten/sycl/GridSamplerKernels.h [hunk 1]
```diff
@@ -0,0 +1,18 @@
+#pragma once
+
+#include <ATen/ATen.h>
+
+namespace at {
+namespace native {
+namespace xpu {
```
Comment (fengyuan14) [inline src/aten/sycl/GridSamplerKernels.h:7]:
```suggestion
namespace at::native::xpu {
```

File: src/aten/sycl/GridSamplerKernels.h [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+Tensor grid_sampler_2d_kernel(
+    const Tensor& input,
+    const Tensor& grid,
+    int64_t interpolation_mode,
+    int64_t padding_mode,
+    bool align_corners);
+
+}
+} // namespace native
+} // namespace at
```
Comment (fengyuan14) [inline src/aten/sycl/GridSamplerKernels.h:18]:
```suggestion
} // namespace at::native::xpu
```

File: src/aten/sycl/UpSample.h
Comment (fengyuan14) [inline src/aten/sycl/UpSample.h:1]: Should reuse PyTorch heard,
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/UpSample.h

File: src/aten/sycl/GridSampler.cpp [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+  index_t out_sC;
+  index_t out_sH;
+  index_t out_sW;
+  scalar_t* grid_data;
+  scalar_t* input_data;
+  scalar_t* output_data;
+};
+
+template <typename scalar_t, typename index_t>
+void grid_sampler_2d_forward_kernel(
```
Comment (fengyuan14) [inline src/aten/sycl/GridSampler.cpp:247]: Renaming, `grid_sampler_2d_forward_template`

File: src/aten/sycl/GridSampler.h [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    return span - extra + min;
+  }
+}
+
+template <typename scalar_t>
+static inline scalar_t safe_downgrade_to_int_range(scalar_t x) {
+  // -100.0 does not have special meaning. This is just to make sure
+  // it's not within_bounds_2d or within_bounds_3d, and does not cause
+  // undefined behavior.
+  if (x > INT_MAX - 1 || x < INT_MIN || !std::isfinite(static_cast<double>(x)))
```
Comment (dvrogozh) [inline src/aten/sycl/GridSampler.h:79]: from compiler:
```
warning: implicit conversion from 'int' to 'float' changes value from 2147483646 to 2147483648 [-Wimplicit-const-int-float-conversion]
   79 |   if (x > INT_MAX - 1 || x < INT_MIN || !std::isfinite(static_cast<double>(x)))
      |         ~ ~~~~~~~~^~~
```

please, resolve.

File: src/aten/sycl/GridSampler.h
Comment (majing921201) [reply src/aten/sycl/GridSampler.h:1]: Many cpu specific head files in [https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/GridSampler.h](url). so we still keep this file
```
#include <ATen/cpu/vec/functional.h>
#include <ATen/cpu/vec/vec.h>
#include <ATen/native/DispatchStub.h>
#include <ATen/native/cpu/utils.h>
```

**Comment by majing921201** [comment]
half/bfloat16 test case enabled:
image

**Comment by majing921201** [comment]
https://github.com/intel/torch-xpu-ops/pull/396 covers both forward and backward

**Pull request closed without merging**